File: test_api.nim
import httpclient, json, os, strutils

const BASE_URL = "http://localhost:8080/api/v1"

proc testHealthCheck() =
  echo "Testing health check endpoint..."
  try:
    let client = newHttpClient()
    let response = client.getContent(BASE_URL & "/health")
    let jsonResponse = parseJson(response)
    
    if jsonResponse{"status"}.getStr() == "ok":
      echo "✓ Health check passed"
    else:
      echo "✗ Health check failed: ", response
  except:
    echo "✗ Health check failed: ", getCurrentExceptionMsg()

proc testGffToJson() =
  echo "\nTesting GFF to JSON conversion..."
  
  # Create a test GFF file (using our shared module)
  let testGffContent = """GFF V1.0
Test: Hello World
Version: 1
"""
  
  try:
    # Write test file
    writeFile("test.gff", testGffContent)
    
    let client = newHttpClient()
    let response = client.postContent(
      BASE_URL & "/convert/gff-to-json",
      multipart = {
        "file": ("test.gff", "application/octet-stream", readFile("test.gff"))
      }
    )
    
    let jsonResponse = parseJson(response)
    if jsonResponse.hasKey("error"):
      echo "✗ GFF to JSON conversion failed: ", jsonResponse{"error"}.getStr()
    else:
      echo "✓ GFF to JSON conversion successful"
      echo "  Response: ", response[0..min(100, response.len-1)], "..."
    
    # Clean up
    removeFile("test.gff")
  except:
    echo "✗ GFF to JSON conversion failed: ", getCurrentExceptionMsg()

proc testJsonToGff() =
  echo "\nTesting JSON to GFF conversion..."
  
  # Create a test JSON file
  let testJsonContent = """{
  "Test": "Hello World",
  "Version": 1
}"""
  
  try:
    # Write test file
    writeFile("test.json", testJsonContent)
    
    let client = newHttpClient()
    let response = client.postContent(
      BASE_URL & "/convert/json-to-gff",
      multipart = {
        "file": ("test.json", "application/json", readFile("test.json"))
      }
    )
    
    if response.len > 0:
      echo "✓ JSON to GFF conversion successful"
      echo "  Response size: ", response.len, " bytes"
      # Save the converted file for inspection
      writeFile("converted.gff", response)
      echo "  Converted file saved as: converted.gff"
    else:
      echo "✗ JSON to GFF conversion returned empty response"
    
    # Clean up
    removeFile("test.json")
  except:
    echo "✗ JSON to GFF conversion failed: ", getCurrentExceptionMsg()

proc testErrorHandling() =
  echo "\nTesting error handling..."
  
  try:
    let client = newHttpClient()
    
    # Test with invalid file format
    let response = client.postContent(
      BASE_URL & "/convert/gff-to-json",
      multipart = {
        "file": ("test.txt", "text/plain", "invalid content")
      }
    )
    
    let jsonResponse = parseJson(response)
    if jsonResponse.hasKey("error") and jsonResponse{"status"}.getInt() == 400:
      echo "✓ Invalid file format error handled correctly"
    else:
      echo "✗ Error handling test failed"
  except:
    echo "✗ Error handling test failed: ", getCurrentExceptionMsg()

proc testRootEndpoint() =
  echo "\nTesting root endpoint..."
  try:
    let client = newHttpClient()
    let response = client.getContent("http://localhost:8080/")
    let jsonResponse = parseJson(response)
    
    if jsonResponse{"service"}.getStr() == "nwn-gff-api":
      echo "✓ Root endpoint working"
    else:
      echo "✗ Root endpoint test failed"
  except:
    echo "✗ Root endpoint test failed: ", getCurrentExceptionMsg()

proc runAllTests() =
  echo "=== NWN GFF API Test Suite ==="
  echo "Make sure the API server is running on port 8080"
  echo "Run: nim c -r gff_api.nim"
  echo ""
  
  # Run tests
  testRootEndpoint()
  testHealthCheck()
  testGffToJson()
  testJsonToGff()
  testErrorHandling()
  
  echo "\n=== Test Suite Complete ==="

when isMainModule:
  runAllTests()

File: tests/test_api.py
"""API endpoint tests"""
import pytest
from fastapi.testclient import TestClient
from app.main import app


client = TestClient(app)


def test_health_check():
    """Test health check endpoint"""
    response = client.get("/api/v1/health")
    assert response.status_code == 200
    data = response.json()
    assert data["status"] == "ok"
    assert data["service"] == "nwn-gff-api"
    assert data["version"] == "0.1.0"


def test_api_info():
    """Test API info endpoint"""
    response = client.get("/api/v1/")
    assert response.status_code == 200
    data = response.json()
    assert data["message"] == "NWN GFF API Service"
    assert data["version"] == "0.1.0"
    assert "endpoints" in data


def test_gff_to_json_invalid_format():
    """Test GFF to JSON with invalid file format"""
    response = client.post(
        "/api/v1/convert/gff-to-json",
        files={"file": ("test.txt", b"invalid content", "text/plain")}
    )
    assert response.status_code == 400
    data = response.json()
    assert "Invalid file format" in data["detail"]


def test_json_to_gff_invalid_format():
    """Test JSON to GFF with invalid file format"""
    response = client.post(
        "/api/v1/convert/json-to-gff",
        files={"file": ("test.txt", b"invalid content", "text/plain")}
    )
    assert response.status_code == 400
    data = response.json()
    assert "Invalid file format" in data["detail"]


def test_sqlite_embed_missing_files():
    """Test SQLite embedding with missing files"""
    response = client.post("/api/v1/convert/sqlite-embed")
    assert response.status_code == 422  # FastAPI validation error


def test_sqlite_extract_invalid_format():
    """Test SQLite extraction with invalid file format"""
    response = client.post(
        "/api/v1/convert/sqlite-extract",
        files={"file": ("test.txt", b"invalid content", "text/plain")}
    )
    assert response.status_code == 400
    data = response.json()
    assert "Invalid file format" in data["detail"]


def test_gff_to_json_valid_gff():
    """Test GFF to JSON conversion with valid GFF-like content"""
    # Create a simple GFF-like content for testing
    gff_content = b"GFF V1.0\nTest: Hello World\nVersion: 1\n"
    
    response = client.post(
        "/api/v1/convert/gff-to-json",
        files={"file": ("test.gff", gff_content, "application/octet-stream")}
    )
    
    # Should succeed with stub implementation
    assert response.status_code == 200
    data = response.json()
    assert "Test" in data
    assert "Version" in data


def test_json_to_gff_valid_json():
    """Test JSON to GFF conversion with valid JSON"""
    json_content = b'{"Test": "Hello World", "Version": 1}'
    
    response = client.post(
        "/api/v1/convert/json-to-gff",
        files={"file": ("test.json", json_content, "application/json")}
    )
    
    # Should succeed and return binary data
    assert response.status_code == 200
    assert response.headers["content-type"] == "application/octet-stream"
    assert "converted.gff" in response.headers["content-disposition"]